File: tutorial/01_01.cpp
//01_01.cpp

#include "rxcpp/rx.hpp"
#include "node_util/node_util.hpp"

using namespace node_util; // console, "string"s + 
namespace rx = rxcpp;
using evt_t  = int;

void test_01_01() {
//-----------------------------------------------------------------------------
  auto on_subscribe = [](auto& observer) {
    observer.on_next(evt_t(11));
    observer.on_completed();
    observer.on_next(evt_t(12));

    auto on_dispose = [] { console::log("on_dispose"); };
    observer.add(on_dispose);
  };

  auto handler_next      = [](evt_t i) { console::log  ("on_next: "s  + i); };
  auto handler_error     = [](auto  e) { console::error("on_error: "s + e); };
  auto handler_completed = []          { console::log  ("on_completed");    };

  //--------------
  auto observable   = rx::observable<>::create<evt_t>(on_subscribe);
  auto subscription = observable.subscribe(handler_next,
                                           handler_error,
                                           handler_completed);
//-----------------------------------------------------------------------------
}

/* 실행 결과

[14:38:28] [LOG]   ----- test start -----
[14:38:28] [LOG]   on_next: 11
[14:38:28] [LOG]   on_completed
[14:38:28] [LOG]   on_dispose
[14:38:28] [LOG]   ----- test end    -----

*/

File: tutorial/01_01.js
//01_01.js

var Rx = require("rx") 






function test_01_01() {
//-----------------------------------------------------------------------------
  function onSubscribe(observer) {
    observer.onNext(11);
    observer.onCompleted();
    observer.onNext(12);

    var onDispose = ()=>console.log("onDispose");
    return onDispose;
  };
  
  var handlerNext      =  i=> console.log  ("onNext: "  + i);
  var handlerError     =  e=> console.error("onError: " + e.message);
  var handlerCompleted = ()=> console.log  ("onCompleted");
  
  //--------------
  var observable   = Rx.Observable.create(onSubscribe);
  var subscription = observable.subscribe(handlerNext,
										                      handlerError,
										                      handlerCompleted);
//-----------------------------------------------------------------------------
}

/* 실행 결과

[14:46:09] [LOG] ----- test start -----
[14:46:09] [LOG] onNext: 11
[14:46:09] [LOG] onCompleted
[14:46:09] [INFO] onDispose
[14:46:09] [LOG] ----- test end   -----

 */
module.exports = test_01_01; 